### PR TITLE
Use dynamic configuration to new Memory pool utility

### PR DESCRIPTION
**Pull request by Efrat1**
Use dynamic configuration to new Memory pool utility

### CODE REVIEW

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 1]
```diff
@@ -374,6 +374,17 @@ bool PreProcessor::validateSubpoolsConfig() {
   return true;
 }
 
+const concordUtil::MultiSizeBufferPool::SubpoolsConfig PreProcessor::defaultSubpoolsConfig_ = {{65536, 0, 1400},
```
Comment (glevkovich) [inline bftengine/src/preprocessor/PreProcessor.cpp]: - here we dont need the explciit numbers, use shift:
e.g (with int, choose any suitable type)
`constexper int MB = (1 << 20);`
so for 1st raw:
{{65 * MB, 0, 1400}

- in below line you used 140 and not 1400
- try defining it as static constexpr  concordUtil::MultiSizeBufferPool::SubpoolsConfig and see if it's compile. if not, just use static const.

File: bftengine/src/preprocessor/PreProcessor.hpp [hunk 2]
```diff
@@ -342,7 +342,9 @@ class PreProcessor {
   // engine supported size is kMaxSubpoolBuffersize_ - responseInternalOverhead_
   static constexpr uint32_t kMaxSubpoolBuffersize_{1UL << 25};
   // Must be sorted in an ascending order, by bufferSize
-  const std::vector<concordUtil::MultiSizeBufferPool::SubpoolConfig> subpoolsConfig_;
+  const concordUtil::MultiSizeBufferPool::SubpoolsConfig dynamicSubpoolsConfig_;
```
Comment (glevkovich) [inline bftengine/src/preprocessor/PreProcessor.hpp]: Why are you keeping it as a member? you are using it only in ctor

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 3]
```diff
@@ -394,46 +405,37 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
...
-          {1 << 19, 0, 1400},  // 512KB
-          {1 << 20, 0, 1400},  // 1MB
-          {1 << 21, 0, 1400},  // 2MB
-          {1 << 22, 0, 700},   // 4MB
-          {1 << 23, 0, 700},   // 8MB
-          {1 << 24, 0, 80},    // 16MB
-          {1 << 25, 0, 80},    // 32MB
-      },
-      // TODO - this is a temporary in-place configuration - should be replaced with a future task to take it from
+      dynamicSubpoolsConfig_{YAML::Load(myReplica_.getReplicaConfig().get<std::string>(
```
Comment (glevkovich) [inline bftengine/src/preprocessor/PreProcessor.cpp]: i dont think we need to initialize this member or have it as member at all.
define a non member function above this one called
`static const concordUtil::MultiSizeBufferPool::SubpoolsConfig calcSubpoolsConfig()`

please read the dynamicSubpoolsConfig_ there, decode it and return the config from replica config or default config.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 4]
```diff
@@ -394,46 +405,37 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
...
-                      : nullptr},
+      bufferPool_{
+          myReplica_.getReplicaConfig().enablePreProcessorMemoryPool
+              ? std::make_unique<concordUtil::MultiSizeBufferPool>(
+                    "PreProcessor::memoryPool",
+                    timers_,
+                    subpoolsConfig_,
+                    concordUtil::MultiSizeBufferPool::Config{
+                        myReplica_.getReplicaConfig().get<uint64_t>(
+                            "concord.preprocessor.multiSizedPool.maxAllocatedBytes", 10ULL * (1 << 30)),
```
Comment (glevkovich) [inline bftengine/src/preprocessor/PreProcessor.cpp]: We didnt talk about real config.
please keep all the same but set maxAllocatedBytesto 10GB and enableStatusReportChangesOnly=false

File: bftengine/src/preprocessor/PreProcessor.hpp [hunk 5]
```diff
@@ -342,7 +342,9 @@ class PreProcessor {
...
+  const concordUtil::MultiSizeBufferPool::SubpoolsConfig subpoolsConfig_;
+  static const concordUtil::MultiSizeBufferPool::SubpoolsConfig defaultSubpoolsConfig_;
```
Comment (glevkovich) [inline bftengine/src/preprocessor/PreProcessor.hpp]: i dont think we need it as member. define it in the suggested calculation fucntion.

**Review by glevkovich — CHANGES_REQUESTED** [review]

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 1 above)
Comment (Efrat1) [reply bftengine/src/preprocessor/PreProcessor.cpp]: OK.
static constexpr is not compiled.

File: bftengine/src/preprocessor/PreProcessor.hpp (hunk 2 above)
Comment (Efrat1) [reply bftengine/src/preprocessor/PreProcessor.hpp]: Removed

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 3 above)
Comment (Efrat1) [reply bftengine/src/preprocessor/PreProcessor.cpp]: Done

File: bftengine/src/preprocessor/PreProcessor.hpp (hunk 5 above)
Comment (Efrat1) [reply bftengine/src/preprocessor/PreProcessor.hpp]: Done

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 4 above)
Comment (Efrat1) [reply bftengine/src/preprocessor/PreProcessor.cpp]: Fixed enableStatusReportChangesOnly,
maxAllocatedBytesto is already 10GB...

**Pull request merged**